### DISCUSSION TITLE
Build Error re: key44 true in remotecontrol.h

### DISCUSSION THREAD

**Discussion by stretch911** [PlummersSoftwareLLC/NightDriverStrip#200] [General]
I have several led remotes laying around, 24 keys and 44 keys. 

I have ENABLE_REMOTE  defined as 1.

Things compiled fine and uploaded to an esp32 wroom 32 (mini). board=esp32dev.
I wanted to try a 44 key remote, so on line 46 and 47 (remotecontrol.h) in latest build, I set key24 to false and key44 to true.

Build Error below:

In file included from src/drawing.cpp:36:
include/remotecontrol.h: In member function 'void RemoteControl::handle()':
include/remotecontrol.h:195:31: error: 'IR_BMINUS' was not declared in this scope
             if (lastResult == IR_BMINUS)
                               ^~~~~~~~~
include/remotecontrol.h:195:31: note: suggested alternative: 'IR_BMinus'
             if (lastResult == IR_BMINUS)
                               ^~~~~~~~~
                               IR_BMinus
include/remotecontrol.h:201:36: error: 'IR_BPLUS' was not declared in this scope
             else if (lastResult == IR_BPLUS)
                                    ^~~~~~~~
include/remotecontrol.h:201:36: note: suggested alternative: 'IR_BPlus'
             else if (lastResult == IR_BPLUS)
                                    ^~~~~~~~
                                    IR_BPlus
include/remotecontrol.h:222:18: error: 'IR_BPLUS' was not declared in this scope
         else if (IR_BPLUS == result)
                  ^~~~~~~~
include/remotecontrol.h:222:18: note: suggested alternative: 'IR_BPlus'
         else if (IR_BPLUS == result)
                  ^~~~~~~~
                  IR_BPlus
include/remotecontrol.h:229:18: error: 'IR_BMINUS' was not declared in this scope
         else if (IR_BMINUS == result)
                  ^~~~~~~~~
include/remotecontrol.h:229:18: note: suggested alternative: 'IR_BMinus'
         else if (IR_BMINUS == result)
                  ^~~~~~~~~
                  IR_BMinus
include/remotecontrol.h:235:18: error: 'IR_SMOOTH' was not declared in this scope
         else if (IR_SMOOTH == result)
                  ^~~~~~~~~
include/remotecontrol.h:235:18: note: suggested alternative: 'IR_SLOW'
         else if (IR_SMOOTH == result)
                  ^~~~~~~~~
                  IR_SLOW
include/remotecontrol.h:240:18: error: 'IR_STROBE' was not declared in this scope
         else if (IR_STROBE == result)
                  ^~~~~~~~~
include/remotecontrol.h:240:18: note: suggested alternative: 'IR_SLOW'
         else if (IR_STROBE == result)
                  ^~~~~~~~~
                  IR_SLOW
include/remotecontrol.h:248:18: error: 'IR_FADE' was not declared in this scope
         else if (IR_FADE == result)
                  ^~~~~~~
include/remotecontrol.h:248:18: note: suggested alternative: 'IR_FADE7'
         else if (IR_FADE == result)
                  ^~~~~~~
                  IR_FADE7
In file included from include/network.h:30,
                 from src/main.cpp:165:
include/remotecontrol.h: In member function 'void RemoteControl::handle()':
include/remotecontrol.h:195:31: error: 'IR_BMINUS' was not declared in this scope
             if (lastResult == IR_BMINUS)
                               ^~~~~~~~~
include/remotecontrol.h:195:31: note: suggested alternative: 'IR_BMinus'
             if (lastResult == IR_BMINUS)
                               ^~~~~~~~~
                               IR_BMinus
include/remotecontrol.h:201:36: error: 'IR_BPLUS' was not declared in this scope
             else if (lastResult == IR_BPLUS)
                                    ^~~~~~~~
include/remotecontrol.h:201:36: note: suggested alternative: 'IR_BPlus'
             else if (lastResult == IR_BPLUS)
                                    ^~~~~~~~
                                    IR_BPlus
include/remotecontrol.h:222:18: error: 'IR_BPLUS' was not declared in this scope
         else if (IR_BPLUS == result)
                  ^~~~~~~~
include/remotecontrol.h:222:18: note: suggested alternative: 'IR_BPlus'
         else if (IR_BPLUS == result)
                  ^~~~~~~~
                  IR_BPlus
include/remotecontrol.h:229:18: error: 'IR_BMINUS' was not declared in this scope
         else if (IR_BMINUS == result)
                  ^~~~~~~~~
include/remotecontrol.h:229:18: note: suggested alternative: 'IR_BMinus'
         else if (IR_BMINUS == result)
                  ^~~~~~~~~
                  IR_BMinus
include/remotecontrol.h:235:18: error: 'IR_SMOOTH' was not declared in this scope
         else if (IR_SMOOTH == result)
                  ^~~~~~~~~
include/remotecontrol.h:235:18: note: suggested alternative: 'IR_SLOW'
         else if (IR_SMOOTH == result)
                  ^~~~~~~~~
                  IR_SLOW
include/remotecontrol.h:240:18: error: 'IR_STROBE' was not declared in this scope
         else if (IR_STROBE == result)
                  ^~~~~~~~~
include/remotecontrol.h:240:18: note: suggested alternative: 'IR_SLOW'
         else if (IR_STROBE == result)
                  ^~~~~~~~~
                  IR_SLOW
include/remotecontrol.h:248:18: error: 'IR_FADE' was not declared in this scope
         else if (IR_FADE == result)
                  ^~~~~~~
include/remotecontrol.h:248:18: note: suggested alternative: 'IR_FADE7'
         else if (IR_FADE == result)
                  ^~~~~~~
                  IR_FADE7
In file included from include/network.h:30,
                 from src/network.cpp:32:
include/remotecontrol.h: In member function 'void RemoteControl::handle()':
include/remotecontrol.h:195:31: error: 'IR_BMINUS' was not declared in this scope
             if (lastResult == IR_BMINUS)
                               ^~~~~~~~~
include/remotecontrol.h:195:31: note: suggested alternative: 'IR_BMinus'
             if (lastResult == IR_BMINUS)
                               ^~~~~~~~~
                               IR_BMinus
include/remotecontrol.h:201:36: error: 'IR_BPLUS' was not declared in this scope
             else if (lastResult == IR_BPLUS)
                                    ^~~~~~~~
include/remotecontrol.h:201:36: note: suggested alternative: 'IR_BPlus'
             else if (lastResult == IR_BPLUS)
                                    ^~~~~~~~
                                    IR_BPlus
include/remotecontrol.h:222:18: error: 'IR_BPLUS' was not declared in this scope
         else if (IR_BPLUS == result)
                  ^~~~~~~~
include/remotecontrol.h:222:18: note: suggested alternative: 'IR_BPlus'
         else if (IR_BPLUS == result)
                  ^~~~~~~~
                  IR_BPlus
include/remotecontrol.h:229:18: error: 'IR_BMINUS' was not declared in this scope
         else if (IR_BMINUS == result)
                  ^~~~~~~~~
include/remotecontrol.h:229:18: note: suggested alternative: 'IR_BMinus'
         else if (IR_BMINUS == result)
                  ^~~~~~~~~
                  IR_BMinus
include/remotecontrol.h:235:18: error: 'IR_SMOOTH' was not declared in this scope
         else if (IR_SMOOTH == result)
                  ^~~~~~~~~
include/remotecontrol.h:235:18: note: suggested alternative: 'IR_SLOW'
         else if (IR_SMOOTH == result)
                  ^~~~~~~~~
                  IR_SLOW
include/remotecontrol.h:240:18: error: 'IR_STROBE' was not declared in this scope
         else if (IR_STROBE == result)
                  ^~~~~~~~~
include/remotecontrol.h:240:18: note: suggested alternative: 'IR_SLOW'
         else if (IR_STROBE == result)
                  ^~~~~~~~~
                  IR_SLOW
include/remotecontrol.h:248:18: error: 'IR_FADE' was not declared in this scope
         else if (IR_FADE == result)
                  ^~~~~~~
include/remotecontrol.h:248:18: note: suggested alternative: 'IR_FADE7'
         else if (IR_FADE == result)
                  ^~~~~~~
                  IR_FADE7
*** [.pio\build\generic\src\drawing.cpp.o] Error 1
*** [.pio\build\generic\src\main.cpp.o] Error 1
*** [.pio\build\generic\src\network.cpp.o] Error 1
================================================================================================ [FAILED] Took 7.98 seconds ================================================================================================

Environment    Status    Duration
-------------  --------  ------------
generic        FAILED    00:00:07.976

**Comment by prschguy** [comment]
Hello Stretch, have enjoyed your posts. Have also been working on the atomlight for several months. Some things I have been able to fix, some things not. In your case, I did find the problem. As Dave once said, spelling does count. In this case, it's a conflict in capitalization, and terms used. I'm not with the code right now, so working from memory. Above in the defines and below in the functions fix the differences in capitalization. Also fix the conflicts between flash and strobe etc. Doesn't seem to matter what you name them, just have to be the same above as below. While I haven't mapped out the encoding yet, will allow some of those key functions to work. Have been more focused on getting the sound reactive part to work. Thinking my efforts would be easier with the TTGO as the audio already works with that, and display is better than heltec board. Just my 2 cents.

**Comment by stretch911** [comment]
what haven't you been able to fix? I am also working on the audio piece. what sensor/microphone are you using? I ordered 2 max4466s and wasn't satisfied with how they "seemed" to work, so tried a little spectrum analyzer that works on a oled and it looked like both of those were crap...i played with the gain pot on the board but still couldn't get them to be "responsive"...one of them would reflect a signal when I clapped or yelled into it. i think they may be trash...dunno. i did see vumax and vumin in the serial output w/ values. I just ordered INMP441...so we'll see if i can get that working...stay tuned.

**Comment by robertlipe** [comment]
Hey, @stretch911 . I just came across this while looking for something else. Here's the patch I've been adding and reverting (sign) every time I do a branch or merge or update or whtatever. It fixes the mismatches that stop it from compiling.

`diff --git a/include/remotecontrol.h b/include/remotecontrol.h
index 5c9c7d9..f733c9d 100644
--- a/include/remotecontrol.h
+++ b/include/remotecontrol.h
@@ -36,8 +36,8 @@

 #if ENABLE_REMOTE

-#define key24  true
-#define key44  false
+//#define key24  true^M
+#define key44  true^M

 void IRAM_ATTR RemoteLoopEntry(void *);

@@ -69,8 +69,8 @@ void IRAM_ATTR RemoteLoopEntry(void *);
 #endif

 #if key44
-#define IR_BPlus  0xFF3AC5  //
-#define IR_BMinus 0xFFBA45  //
+#define IR_BPLUS  0xFF3AC5  //^M
+#define IR_BMINUS 0xFFBA45  //^M
 #define IR_ON     0xFF827D  //
 #define IR_OFF    0xFF02FD  //
 #define IR_R      0xFF1AE5  //
@@ -111,8 +111,10 @@ void IRAM_ATTR RemoteLoopEntry(void *);
 #define IR_FLASH  0xFFD02F  //
 #define IR_JUMP3  0xFF20DF  //
 #define IR_JUMP7  0xFFA05F  //
-#define IR_FADE3  0xFF609F  //
+#define IR_FADE  0xFF609F  //^M
 #define IR_FADE7  0xFFE01F  //
+#define IR_SMOOTH -1  // 0xFFE01F  //^M
+#define IR_STROBE -1  // 0xFFE01F  //^M
 #endif

 const static struct
@@ -174,4 +176,4 @@ class RemoteControl
     void handle();
 };

-#endif
\ No newline at end of file
+#endif^M

`

I had no idea what "Strobe" or "Smooth" were supposed to even be (not on any of my dozens of remotes...) so I just clubbed them to -1/0xffffffff to keep them out of harms way.

In case you're patching this by hand, it's:
Fix the ase for IR_BPLUS/MINUUS
Use FADE instead of FADE3
Define SMOOTH and STROBE to be something harmless.

This got me effect previous and next, start and stop and all the color keys.


I'll have a big rework in this area, if someone doesn't beat me to it.
